src/net/ossl.cc

```diff
@@ -75,4 +79,98 @@ class ossl_error : public std::system_error {
     }
 };
 
+template<typename T, auto fn>
+struct ssl_deleter {
+    void operator()(T* ptr) { fn(ptr); }
+};
+
+template<typename T, auto fn>
+using ssl_handle = std::unique_ptr<T, ssl_deleter<T, fn>>;
+
+using bio_ptr = ssl_handle<BIO, BIO_free>;
+using evp_pkey_ptr = ssl_handle<EVP_PKEY, EVP_PKEY_free>;
+
+class tls::dh_params::impl {
+public:
+    static int level_to_bits(level l) {
+        switch (l) {
+            case level::LEGACY:
+                return 1776;
+            case level::MEDIUM:
+                return 2432;
+            case level::HIGH:
+                return 3248;
+            case level::ULTRA:
+                return 15424;
+            default:
+                throw std::runtime_error(format("Unknown value of dh_params::level: {:d}", static_cast<std::underlying_type_t<level>>(l)));
+        }
+    }
+
+    static std::shared_ptr<EVP_PKEY> make_evp_pkey(level l) {
+        /// Instantiate new Diffie-Hellman key context
+        EVP_PKEY *pkey = nullptr;
+        EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_from_name(nullptr, "DH", nullptr);
+
+        OSSL_PARAM params[3];
+        unsigned int bits = level_to_bits(l);
+        char group[] = "group";
+        params[0] = OSSL_PARAM_construct_utf8_string("type", group, strlen(group));
```

Comment:
Note: Not sure if this is working as expected, will matter for phase 2 when mTLS is to be implemented